pkg/cloudprovider/providers/vsphere/vsphere.go

```diff
@@ -1242,39 +1244,146 @@ func (vs *VSphere) CreateVolume(volumeOptions *VolumeOptions) (volumePath string
 		return "", err
 	}
 
-	// vmdks will be created inside kubevols directory
-	kubeVolsPath := filepath.Clean(ds.Path(VolDir)) + "/"
-	err = makeDirectoryInDatastore(vs.client, dc, kubeVolsPath, false)
-	if err != nil && err != ErrFileAlreadyExist {
-		glog.Errorf("Cannot create dir %#v. err %s", kubeVolsPath, err)
-		return "", err
+	if volumeOptions.StorageProfileData != "" {
+		// Check if the datastore is VSAN if any capability requirements are specified.
+		// VSphere cloud provider now only supports VSAN capabilities requirements
+		ok, err := checkIfDatastoreTypeIsVSAN(vs.client, ds)
+		if err != nil {
+			return "", fmt.Errorf("Failed while determining whether the datastore: %q"+
+				" is VSAN or not.", datastore)
+		}
+		if !ok {
+			return "", fmt.Errorf("The specified datastore: %q is not a VSAN datastore."+
+				" The policy parameters will work only with VSAN Datastore."+
+				" So, please specify a valid VSAN datastore", datastore)
+		}
 	}
-	glog.V(4).Infof("Created dir with path as %+q", kubeVolsPath)
 
-	vmDiskPath := kubeVolsPath + volumeOptions.Name + ".vmdk"
+	dummyVMName := "dummy-Kubernetes-vm"
+	virtualMachineConfigSpec := types.VirtualMachineConfigSpec{
```

Comment:
Please see the reply for your individual comments.
"Dummy VM is workaround until next VC release" -  You are true.

"We should use it only when absolutely needed i.e. only when vSAN Policy is used." - I don't agree with this. So you are proposing if there is VSAN policy, have dummy VM logic. If no VSAN policy, use the existing virtualDiskManager.CreateVirtualDisk() API.

In next VC release when we FCD's. We can remove the entire which is dummy VM logic and virtualDiskManager.CreateVirtualDisk() logic altogether. At that we can a single function exposed by FCD's which VStorageManager.CreateDisk() API which will accept storagepolicy as an optional parameter. 

If we planning to remove the code in entirety why leave it in chunks is my question over here. Anyways, I am now adding the logic for clean up as well for dummy VM folders.